### ISSUE TITLE
Doxygen Build Error

### ISSUE THREAD

**Issue by Leandros** [mosra/m.css#57]
I must've added something, which broke the doxygen build with `dox2html5.py`:

```
Traceback (most recent call last):
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 3041, in <module>
    run(doxyfile, os.path.abspath(args.templates), args.wildcard, args.index_pages, search_merge_subtrees=not args.search_no_subtree_merging, search_add_lookahead_barriers=not args.search_no_lookahead_barriers, search_merge_prefixes=not args.search_no_prefix_merging)
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 2955, in run
    parsed = parse_xml(state, file)
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 2282, in parse_xml
    typedef = parse_typedef(state, memberdef)
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 1558, in parse_typedef
    typedef.description, templates, search_keywords, typedef.is_deprecated = parse_typedef_desc(state, element)
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 1427, in parse_typedef_desc
    assert not parsed.params and not parsed.return_value and not parsed.return_values and not parsed.exceptions
AssertionError
make: *** [html] Error 1
```

**Comment by mosra** [comment]
Hi :) Without knowing the code that triggered it, it could mean you have a templated `using` declaration and describing the template params with `@param` instead of `@tparam`. If that's not the case, you can try running the script with `--debug`, it'll say what xml file it is processing before it asserts. Paste the file here (or if it's a private project, send it to mosra@centrum.cz). Thanks! :)-------- Původní zpráva --------Předmět: [mosra/m.css] Doxygen Build Error (#57)Odesílatel: Arvid Gerstmann Komu: "mosra/m.css" Kopie: Subscribed I must've added something, which broke the doxygen build with dox2html5.py:
Traceback (most recent call last):
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 3041, in <module>
    run(doxyfile, os.path.abspath(args.templates), args.wildcard, args.index_pages, search_merge_subtrees=not args.search_no_subtree_merging, search_add_lookahead_barriers=not args.search_no_lookahead_barriers, search_merge_prefixes=not args.search_no_prefix_merging)
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 2955, in run
    parsed = parse_xml(state, file)
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 2282, in parse_xml
    typedef = parse_typedef(state, memberdef)
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 1558, in parse_typedef
    typedef.description, templates, search_keywords, typedef.is_deprecated = parse_typedef_desc(state, element)
  File "/Users/arvidgerstmann/p4/depot/extern/bin/shared/m.css/doxygen/dox2html5.py", line 1427, in parse_typedef_desc
    assert not parsed.params and not parsed.return_value and not parsed.return_values and not parsed.exceptions
AssertionError
make: *** [html] Error 1


—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/mosra/m.css","title":"mosra/m.css","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/email/message_cards/header.png","avatar_image_url":"https://assets-cdn.github.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/mosra/m.css"}},"updates":{"snippets":[{"icon":"DESCRIPTION","message":"Doxygen Build Error (#57)"}],"action":{"name":"View Issue","url":"https://github.com/mosra/m.css/issues/57"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/mosra/m.css/issues/57",
"url": "https://github.com/mosra/m.css/issues/57",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
},
{
"@type": "MessageCard",
"@context": "http://schema.org/extensions",
"hideOriginalBody": "false",
"originator": "AF6C5A86-E920-430C-9C59-A73278B5EFEB",
"title": "Doxygen Build Error (#57)",
"sections": [
{
"text": "",
"activityTitle": "**Arvid Gerstmann**",
"activityImage": "https://assets-cdn.github.com/images/email/message_cards/avatar.png",
"activitySubtitle": "@Leandros",
"facts": [
{
"name": "Repository: ",
"value": "mosra/m.css"
},
{
"name": "Issue #: ",
"value": 57
}
]
}
],
"potentialAction": [
{
"name": "Add a comment",
"@type": "ActionCard",
"inputs": [
{
"isMultiLine": true,
"@type": "TextInput",
"id": "IssueComment",
"isRequired": false
}
],
"actions": [
{
"name": "Comment",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueComment\",\n\"repositoryFullName\": \"mosra/m.css\",\n\"issueId\": 57,\n\"IssueComment\": \"{{IssueComment.value}}\"\n}"
}
]
},
{
"name": "Close issue",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueClose\",\n\"repositoryFullName\": \"mosra/m.css\",\n\"issueId\": 57\n}"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/mosra/m.css/issues/57"
}
],
"@type": "OpenUri",
"name": "View on GitHub"
},
{
"name": "Unsubscribe",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"MuteNotification\",\n\"threadId\": 373172598\n}"
}
],
"themeColor": "26292E"
}
]

**Comment by Leandros** [comment]
I tried to figure out what triggered it, to no avail. 
Attached is the full debug log:
[err.tar.gz](https://github.com/mosra/m.css/files/2328472/err.tar.gz)

**Comment by mosra** [comment]
It says the error is somewhere inside `namespacea.xml`. Can you send me that file? (Again, e-mail is fine too.) Thanks!

**Comment by mosra** [comment]
... it should be somewhere in the `xml/` dir inside the output directory.

**Comment by Leandros** [comment]
Done.

**Comment by mosra** [comment]
It looks like `a::PrintfCallback` "is to blame" -- it has a `@return` documentation. If you remove it or make it into a normal paragraph (without `@return`), the generation should work.

I'm putting "is to blame" in quotes, because I'm not sure what to do there -- it's a function pointer, so it kinda might make sense to document its return type (and parameters), but on the other hand Doxygen doesn't help me here at all, doesn't provide separate return type and arguments or anything ... and if there's something I don't want to do, it's parsing C++ ;)

Doxygen itself doesn't care, it'll just create documentation from whatever you write (so you can have `@return` for namespace and it won't compain one bit), but I personally didn't see a point in allowing that, so that's why the asserts.

So there are these options:

- I could make it support `@return` (and `@param`) for function pointer typedefs, but that'll mean I need to parse function pointer definitions, which is a lot of work and a lot can go wrong there. 
- I could issue a warning instead of asserting, saying `@return` can be only used for function documentation (and then ignore the paragraph)
- I could issue a warning and put the contents of the `@return` as a normal paragraph (which might or might not look acceptable)
- I could also keep the current behavior, but you're apparently not alone with this -- #51 is a similar case

**Comment by Leandros** [comment]
I'd be fine already with getting a human readable error message, allowing me to fix my mistake.

**Comment by mosra** [comment]
Which is roughly the second option above. Will do that, then :)

Is it okay if I ship the fix a bit later (Thursday/Friday)? I am a bit late other things right now and need to focus on these.

**Comment by Leandros** [comment]
Take all the time you need, I've been sitting on that error for a couple days already anyway.

**Comment by mosra** [comment]
.. but the suggestion above (removing the `@return` command) did work, right?

**Comment by Leandros** [comment]
Yes, worked perfectly fine.

**Comment by mosra** [comment]
Should be fixed in 394339ea6ed6bcf4158b1f82d5bc5537880bf093. From now on, you'll get warnings instead of hard assertions.
